win32: Windows 10: timeBeginPeriod on demand

avih opened the pull request:
Before this commit, timeBeginPeriod(1) was set once when mpv starts,
and the timers remained hi-res till mpv exits.

Now we do the same as before on Windows version < 10.
On Windows 10+ we now use timeBeginPeriod if needed, per timeout.

To force a mode regardless of Windows version, set env MPV_HRT:
- "always":  the old behavior - hires timers as long as mpv runs.
- "perwait": sets 1ms timer resolution if timeout < 50ms.
- "never":   don't use timeBeginPeriod at all.

It was observed that on Windows 10 we lose about 0.5ms accuracy of
timeouts with "perwait" mode (acceptable), but otherwise it works
well for continuous timeouts (one after the other) and random ones.

On Windows 7 with "perwait": continous timeouts are accurate, but
random timeouts (after some time without timeouts) have bad
accuracy (roughly 16ms resolution, instead of the requested 1ms).

Windows 8 was not tested, so to err on the side of caution, we keep
the legacy behavior "always" by default.

rossy commented on osdep/timer-win2.c:
```diff
@@ -52,7 +81,37 @@ uint64_t mp_raw_time_us(void)
+    }
+
+    // "always"/"never"/"perwait"  (or "auto" - same as unset)
+    if (!(v = getenv("MPV_HRT")) || !strcmp(v, "auto"))
```
Would it make sense for ``MPV_HRT`` and ``MPV_HRT_MAX`` to be combined into a single variable that accepts "auto", "always" and "never" keywords as well as a 1-1000 per-wait value? That would make one less variable to set if the user wants to customize this permanently.

avih replied:
> That would make one less variable to set if the user wants to customize this permanently.

I don't think this should be a consideration, and personally I'm not very fond of a variable which could be either a string or a number.

> single variable that accepts "auto", "always" and "never" keywords or a 1-1000 per-wait value

This would not allow to set the resolution in "always" or "auto" modes.

Also, keep in mind that we might want to enhance `MPV_HRT` soon to also support "playback"  (or "perwait+playback" etc).

rossy replied:
> I don't think this should be a consideration, and personally I'm not very fond of a variable which could be either a string or a number.

Well, mpv is full of such options. It would also make it slightly easier to instruct users to change the value for testing purposes.

> This would not allow to set the resolution in "always" or "auto" modes.

I'm only talking about merging ``MPV_HRT`` and ``MPV_HRT_MAX``. If I understand correctly, that's not the resolution, but the threshold timeout value to use in per-wait mode.

> Also, keep in mind that we might want to enhance MPV_HRT soon to also support "playback" (or "perwait+playback" etc).

This could still be a separate variable (``MPV_HRT_PLAYBACK=yes``.) I think it could also be an mpv option.

rossy replied:
Come to think of it, I don't mind too much either way. I can see us adding more "conditions" for the hires timer in future that might need an integer argument, so if you think they should stay as separate variables, that's fine by me.

avih replied:
> I'm only talking about merging MPV_HRT and MPV_HRT_MAX. If I understand correctly, that's not the resolution, but the threshold timeout value to use in per-wait mode.

You understood correctly, I misread your comment. My bad.

However, you would not be able to set a custom perwait threshold while letting mpv decide between perwait/always depending on the windows version.

> This could still be a separate variable (MPV_HRT_PLAYBACK=yes.) I think it could also be an mpv option.

True for both.

> Come to think of it, I don't mind too much either way.

OK, so mainly due to the reduced functionality (can't describe auto+threshold), I'll stick with the separate values.

If I overlooked something and the functionality is not indeed reduced, then I don't mind much either.

rossy replied:
> However, you would not be able to set a custom perwait threshold while letting mpv decide between perwait/always depending on the windows version.

I don't really see that as a big use-case, since if someone is setting a custom threshold, they probably want per-wait mode. Maybe if it was an mpv option it would make sense, since someone could be sharing the same config file between Win7 and Win10, but I don't see the same thing happening with env vars. Still, I'm fine either way.

avih replied:
>  if someone is setting a custom threshold, they probably want per-wait mode

Agreed, _probably_.

> someone could be sharing the same config file between Win7 and Win10, but I don't see the same thing happening with env vars

Why not? it being env vars rather than option is imposed on the user, if they could do that with at a config file then they would IMO. If I'm sharing a config between my window 7 and 10, including env vars via a batch file wrapper, I wouldn't want to check the windows version myself at the batch file.

> Still, I'm fine either way.

OK, so for the bigger use cases coverage, I'll keep it separate. Thanks.

rossy replied:
> If I'm sharing a config between my window 7 and 10, including env vars via a batch file wrapper, I wouldn't want to check the windows version myself at the batch file.

Yeah, I guess that's a reasonable case for separate vars.

rossy reviewed (changes requested):
Looks good apart from a couple of nits and some questions

avih commented on osdep/timer-win2.c:
```diff
@@ -52,7 +81,37 @@ uint64_t mp_raw_time_us(void)
+    // for easier experimentation and diagnostic of bug reports.
+    const char *v;
+
+    // 1..1000 ms max timtout for hires (only when MPV_HRT is unset or "auto")
```
> Also, this is for "perwait" as well, right?

Yeah, initially "auto" was what now is "perwait". I did notice it too, but didn't push the update yet. Will do :)

avih commented on osdep/timer-win2.c:
```diff
@@ -19,22 +19,51 @@
 
+// ms values
+static int hires_max = 50;
+static int hires_res = 1;
```
> I'm tempted to default this to 4ms. Thoughts?

My main motivation to keep 1ms was to avoid regressions as much as possible (we don't actually know what might be affected, and in what ways, if we reduce the resolution).

But after some time, if it turns out that 4ms is good enough, I think it would be fine to change the default.

rossy replied:
Yep, that's reasonable

avih commented on osdep/timer-win2.c:
```diff
@@ -19,22 +19,51 @@
 {
     if (us < 0)
         return;
+    int hrt = mp_start_hires_timers(us / 1e3);
```
> Otherwise <1000μs waits won't change the timer freq, even though `Sleep(1)` will be called.

They would change the resolution, because 1e3 is double, hence the value would be >0. Nevertherless I'll do the change you request, but it's no-op.

rossy replied:
Right, I didn't realise ``mp_start_hires_timers`` had a double argument. I think it should be an int, since timeBeginPeriod, Sleep and SleepConditionVariable* all take an integer argument.

avih replied:
> Right, I didn't realise `mp_start_hires_timers` had a double argument. I think it should be an int, since timeBeginPeriod, Sleep and SleepConditionVariable* all take an integer argument.

I intentionally made it a double so that any value bigger than 0 (in any timeout resolution, including sub-ms or even sub-us) would trigger hires. This is pretty much in line with the logic at `mp_sleep_us` for sub-ms values (and the reason why the change you requested initially is almost no-op), and I still think it's the correct thing to do.

Agreed?

rossy replied:
Well, to be in line with the mpv functions, it would have to take int64 microseconds. As a double it's not in line with Windows conventions or mpv conventions. Since we have to convert the timeout to milliseconds for Windows anyway (including rounding up the timeout value to 1ms,) I think it makes sense to pass the exact same integer value to ``mp_start_hires_timers`` that we pass to the Windows API.

avih replied:
> to be in line with the mpv functions, it would have to take int64 microseconds. As a double it's not in line with Windows conventions or mpv conventions

I still think double is the most generic and native way to express a value which can then be used with logic of "anything which isn't exactly 0 should get hires timers" regardless of platform or application conventions, however, I also don't think it's worth spending more time on this, and at least currently, using microsec argument doesn't reduce its usability.

I'll prefer it int64 microsec value to be aligned with the mpv conventions. I don't think it should align with the windows conventions, because the timeout value comes from mpv, and that's what this function should take into account in this context - regardless of the value which ends up being used with the windows API.

Are you OK with int64 microsec?

rossy replied:
I'll admit that part of this is because I don't like unnecessary use of floating point arithmetic (even though I know there aren't accuracy problems in this case) so I'm okay with int64 microseconds.

I'm not sure why you're against integer milliseconds though. Both call sites have an integer millisecond value easily available, and the other values that ``mp_start_hires_timers`` uses (``hires_max`` and ``hires_res``) are also integer milliseconds, so there would be less conversion involved if ``wait_ms`` was integer milliseconds too. ``mp_start_hires_timers`` isn't a public API, so we could always change it to use microseconds if we need to in future.

avih replied:
> I'm not sure why you're against integer milliseconds though

Because it can't represent the concept which the logic applies - anything which isn't exactly 0, e.g. even one nanosec.

If 0 was included in the range of hires, then it wouldn't matter to me.

avih replied:
Changed to int.

avih commented:
- Addressed the comments for the "timeout" typo and change+move the 1e3 division.
- Minor comments changes at the MPV_HRT processing.
- Changed one condition from `>0` to `>=1` for clarity (int, no-op).
- Changed upper perwait threshold to inclusive for consistent terminology (`<=50`).

Still awaiting conclusion (didn't change yet):
- 1ms to 4ms
- Unification of the env MPV_HRT.

avih commented:
Force pushed: changed double to int `wait_ms` argument `int mp_start_hires_timers(int wait_ms)`.

rossy commented:
LGTM

avih commented:
Thanks. And may the force be with us.
